src/overlays/actors/ovl_Demo_Du/z_demo_du.c

```diff
@@ -1,228 +1,1044 @@
+extern AnimationHeader D_06005458;
+extern AnimationHeader D_06006104;
+extern AnimationHeader D_060067CC;
+extern AnimationHeader D_06006EB0;
 extern UNK_TYPE D_06007FC0;
-extern UNK_TYPE D_06011CA8;
-extern UNK_TYPE D_06012014;
-
-/*
-const ActorInit Demo_Du_InitVars = {
-    ACTOR_DEMO_DU,
-    ACTORCAT_NPC,
-    FLAGS,
-    OBJECT_DU,
-    sizeof(DemoDu),
-    (ActorFunc)DemoDu_Init,
-    (ActorFunc)DemoDu_Destroy,
-    (ActorFunc)DemoDu_Update,
-    (ActorFunc)DemoDu_Draw,
+extern FlexSkeletonHeader D_06011CA8;
+extern AnimationHeader D_06012014;
+
+static s32 sUnused = 0;
+
+#include "z_demo_du_cutscene_data.c" EARLY
+
+static Gfx* sEyeTextures[] = { 0x06008080, 0x06008480, 0x06008880, 0x0600A540 };
+static Gfx* sMouthTextures[] = { 0x06008C80, 0x06009D40, 0x0600A940, 0x0600B180 };
+
+/**
+ * Cs => Cutscene
+ *
+ * FM => Fire Medallion
+ * GR => Goron's Ruby
+ * AG => In the chamber of sages, just After the final blow on Ganon.
+ * CR => Credits
+ *
+ */
+
+// Each macro maps its argument to an index of sUpdateFuncs.
+#define CS_FIREMEDALLION_SUBSCENE(x) (0 + (x))      // DEMO_DU_CS_FIREMEDALLION
+#define CS_GORONSRUBY_SUBSCENE(x) (7 + (x))         // DEMO_DU_CS_GORONS_RUBY
+#define CS_CHAMBERAFTERGANON_SUBSCENE(x) (21 + (x)) // DEMO_DU_CS_CHAMBER_AFTER_GANON
+#define CS_CREDITS_SUBSCENE(x) (24 + (x))           // DEMO_DU_CS_CREDITS
+
+void DemoDu_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    DemoDu* this = THIS;
+
+    SkelAnime_Free(&this->skelAnime, globalCtx);
+}
+
+void DemoDu_UpdateEyes(DemoDu* this) {
+    s16* blinkTimer = &this->blinkTimer;
+    s16* eyeTexIndex = &this->eyeTexIndex;
+    s32 pad[3];
+
+    if (DECR(*blinkTimer) == 0) {
+        *blinkTimer = Rand_S16Offset(60, 60);
+    }
+
+    *eyeTexIndex = *blinkTimer;
+    if (*eyeTexIndex >= 3) {
+        *eyeTexIndex = 0;
+    }
+}
+
+void DemoDu_SetEyeTexIndex(DemoDu* this, s16 eyeTexIndex) {
+    this->eyeTexIndex = eyeTexIndex;
+}
+
+void DemoDu_SetMouthTexIndex(DemoDu* this, s16 mouthTexIndex) {
+    this->mouthTexIndex = mouthTexIndex;
+}
+
+// Resets all the values used in this cutscene.
+void DemoDu_CsAfterGanon_Reset(DemoDu* this) {
+    this->updateIndex = CS_CHAMBERAFTERGANON_SUBSCENE(0);
+    this->drawIndex = 0;
+    this->shadowAlpha = 0;
+    this->demo6KSpawned = 0;
+    this->actor.shape.shadowAlpha = 0;
+    this->unk_1A4 = 0.0f;
+}
+
+void DemoDu_CsAfterGanon_CheckIfShouldReset(DemoDu* this, GlobalContext* globalCtx) {
+    static s32 D_8096CE94 = false;
+
+    if (globalCtx->csCtx.state == 0) {
+        if (D_8096CE94) {
+            if (this->actor.params == DEMO_DU_CS_CHAMBER_AFTER_GANON) {
+                DemoDu_CsAfterGanon_Reset(this);
+            }
+            D_8096CE94 = false;
+            return;
+        }
+    } else if (!D_8096CE94) {
+        D_8096CE94 = true;
+    }
+}
+
+// Returns true if the animation finished
+s32 DemoDu_UpdateSkelAnime(DemoDu* this) {
+    return SkelAnime_Update(&this->skelAnime);
+}
```

Comment:
non-looping animations specifically.
I do think that this comment is better left for the `SkelAnime_Update` function though